### ISSUE TITLE
Possibility to cache to RAM ?

### ISSUE THREAD

**Issue by knarf-hackatal2016** [joblib/joblib#527]
The memory module enable caching to disk. However in some situations (especially with scikit-learn's Pipelines) it would also be desirable to have the option to cache to RAM for faster cache/load.

**Comment by GaelVaroquaux** [comment]
> The memory module enable caching to disk. However in some situations
> (especially with scikit-learn's Pipelines) it would also be desirable to have
> the option to cache to RAM for faster cache/load.

Agreed. A pull request would be fantastic (we would have to iterate on
the right strategy).

In the mean time, a strategy is to use /dev/shm under Linux, which is
in-memory.

**Comment by knarf-hackatal2016** [comment]
I'd like to submit it as a PR as soon as I have the time to format it better. What I have now is readable but doesnt fit nicely inside the memory module. (it's two classes that inherit from Memory and MemorizedFunc and rewrite main methods). Should I submit that?(faster)

**Comment by lesteve** [comment]
You can always open a PR with your code as it is now so we can have a quick look.

Longer-term, I guess this should fit with @aabadie's work of Memory storage, see https://github.com/joblib/joblib/pull/397.

**Comment by FranckCharras** [comment]
Sorry I didn't submit any PR because my implementation didn't seem to work reliably (multiple jobs spawn multiple caches...) and I couldn't invest much time in fixing it yet. I want now to investigate the possibility of using [http://docs.pyfilesystem.org/en/latest/reference/memoryfs.html](pyfilesystem), which seems to work with all platforms, to simply write the pickles to RAM.

**Comment by aabadie** [comment]
If you are using Linux, you could take advantage of tmpfs in `/dev/shm` with something like this:
```python
mem = Memory("/dev/shm/joblib")

mem.cache(your_func)
```
This will cache into memory since tmpfs is an in-memory filesystem and without the multiprocess issue.

This is just a suggestion I didn't it myself.

**Comment by fcharras** [comment]
The problem encountered with multiprocessing seems completely solved when using multiprocessing.Manager.dict() to share the cache between all processes so I may finally be able to submit a working PR, which works on all OSes. The performance gain should be especially noticeable when using a sklearn's Pipeline with many cached steps.

I don't want to use the workaround with /dev/shm because it's not portable, and also I suspect it is not optimal because data that is stored there will be copied elsewhere in RAM upon reading by each process, hence will be slower and consume far more memory than necessary.

(on a side note, using the memoryfs from pyfilesystem is a portable solution, but would have the same performance drawback than when using /dev/shm)

**Comment by aabadie** [comment]
Thanks for looking further into this.

Maybe this #397 could help you: you could add a specific store backend where object are stored in memory.

(just pushing again this PR to be reviewed ;))

**Comment by fcharras** [comment]
Alright here is what I've been using for caching to RAM:
https://github.com/joblib/joblib/pull/534

**Comment by fcharras** [comment]
The previous script https://github.com/joblib/joblib/pull/534 seems to run as expected but I don't think the performance is great. Moreover the manager.dict() seems to fail with very large objects (maybe because of pickling ?) and it makes processes hang.

After further investigation, the bottleneck for my use seems to comes from unpickling large objects rather than read/write operations. I'd like to be able to get multiple threads to read the cache at the same time without pickling. This is a harder problem as I expected (and seems impossible in python).

However having a shared store of pickled objects in RAM seems possible.

**Comment by fcharras** [comment]
A little update on this because I think I've found another way.

My goal was to perform a grid search on a Pipeline object with n queued estimators with minimum computation overhead. I don't want to compute each step with a given set of parameters more than once, so caching is good, but disks read/write and pickling is too long, so I'd just like to keep everything in RAM.

This could actually be done by nesting GridSearch. Let's assume A,B are transformers and C is a classifier, with sets of possible parameters Sa, Sb and Sc. We want to find the best parameters in Sa, Sb and Sc such that the pipeline object Pipeline(A,B,C) has the best score. Then the following sklearn object:
```
GridSearchCV( 
    Pipeline(A,
        GridSearchCV(
            Pipeline(B,
                GridSearchCV( C, params=Sc )
            ), 
        params=Sb )
    ), params=Sa
)
would do it, and would keep every intermediate result in RAM until it's not needed anymore. Now it's not pretty and uneasy to get the results of the GridSearch but I think it would work (haven't tested it yet though).

**Comment by dermoth** [comment]
> I don't want to use the workaround with /dev/shm because it's not portable, and also I suspect it is not optimal because data that is stored there will be copied elsewhere in RAM upon reading by each process, hence will be slower and consume far more memory than necessary.

FYI I haven't yet tried this module (I was just looking for the right one to use...) however I'd like to point out if you could make the cache mmap data on disk then the cache wouldn't consume much more ram with multiple threads/processes using /dev/shm. Only the actual data fetched from cache would have to be copied. Even better, using regular on-disk files, the os would actually use your available ram as cache shared between threads/processes.

Ideally you needs a thread-safe storage engine, and one just happen to exist: see lmdb (there's even a python binding)... So maybe adding an lmdb backend could be a more sensible solution to solve this issue?

**Comment by skjerns** [comment]
> If you are using Linux, you could take advantage of tmpfs in `/dev/shm` with something like this:
> 
> ```python
> mem = Memory("/dev/shm/joblib")
> 
> mem.cache(your_func)
> ```
> 
> This will cache into memory since tmpfs is an in-memory filesystem and without the multiprocess issue.
> 
> This is just a suggestion I didn't try it myself.

Want to highlight once more that this isn't a good or performant approach. In my case storing was slightly faster (expected), but retrieval slower. Is there any other way to implement a RAM backend?

```Python
import os
import numpy as np
from joblib import Memory
import stimer

mem1 = Memory('/dev/shm/joblib', verbose=0)
mem2 = Memory(os.path.expanduser('~/Desktop/'), verbose=0)

@mem1.cache()
def test1():
    return np.ones([1000, 1000, 1000], dtype=float)

@mem2.cache()
def test2():
    return np.ones([1000, 1000, 1000], dtype=float)


with stimer('storing shm'):
    test1()

with stimer('storing disk'):
    test2()

with stimer('retrieving shm'):
    test1()

with stimer('retrieving disk'):
    test2()



Elapsed storing shm: 3.8 sec
Elapsed storing disk: 5.2 sec
Elapsed retrieving shm: 2.3 sec
Elapsed retrieving disk: 2.0 sec
```

**Comment by fcharras** [comment]
This is an old thread :grin: . When I started it I had little understanding of what was really going on.

It turns out naively trying to have a cache in ram for a multi-processed job will not work since different processes have separate memory spaces so they can't just write in a dict in the main process. I'm not sure I was well aware of that at the time :grin: Also, overhead in IPC (inter process communication) can happen either at serialization step or unserialization and there might be more to it than IO bandwith in source and target. Likely profiling and tweaking at this level is not trivial and it's not a given that using ram for IPC will noticeably improve performance.

That said, nowadays I'd suggest to look at joblib.Memory `backend` [abstractions](https://github.com/joblib/joblib/blob/master/joblib/_store_backends.py#L39) and maybe try to plug into that the new [multiprocessing.shared_memory](https://docs.python.org/3/library/multiprocessing.shared_memory.html) module. (it's "new" from the perspective of this issue, but in fact it's been introduced in python 3.8)

Should enable similar performance than writing into `/dev/shm` without the issues that come with it (portability / cleaning / limited space / etc)

I think recently someone told me it's actually possible and it can actually improve some workloads so maybe a PR in this direction would be interesting.
